exercises/chapter12/test/Main.purs

```diff
@@ -1,11 +1,148 @@
+      test "missing" do
+        chars <- countCharacters $ fp [ inDir, "foof.txt" ]
+        Assert.equal (Left "ENOENT: no such file or directory, open 'test/data/foof.txt'") $ lmap message chars
+    test "writeGet" do
+      let
+        outFile = fp [ outDir, "user.txt" ]
+
+        expectedOutFile = fp [ inDir, "user.txt" ]
+      writeGet "https://reqres.in/api/users/1" outFile
+      -- Check for valid write
+      actualOutTxt <- readTextFile UTF8 outFile
+      expectedOutTxt <- readTextFile UTF8 expectedOutFile
+      Assert.equal expectedOutTxt actualOutTxt
+    test "concatenateManyParallel" do
+      let
+        inFiles = map (\i -> fp [ inDir, "many", "file" <> show i <> ".txt" ]) $ 1 .. 9
+
+        outFile = fp [ outDir, "manyConcatParallel.txt" ]
+
+        expectedOutFile = fp [ inDir, "manyConcat.txt" ]
+      concatenateManyParallel inFiles outFile
+      -- Check for valid concat
+      actualOutTxt <- readTextFile UTF8 outFile
+      expectedOutTxt <- readTextFile UTF8 expectedOutFile
+      Assert.equal expectedOutTxt actualOutTxt
+    suite "getWithTimeout" do
+      test "valid site" do
+        let
+          expectedOutFile = fp [ inDir, "user.txt" ]
+        actual <- getWithTimeout 1000.0 "https://reqres.in/api/users/1"
+        expected <- Just <$> readTextFile UTF8 expectedOutFile
+        Assert.equal expected actual
+      test "no response" do
+        actual <- getWithTimeout 10.0 "https://example.com:81"
+        Assert.equal Nothing actual
+    suite "recurseFiles" do
+      let
+        recurseDir = fp [ inDir, "tree" ]
+      test "many files" do
+        expectedTxt <- readTextFile UTF8 $ fp [ recurseDir, "expected.txt" ]
+        let
+          expected = Path.normalize <$> split (Pattern "\n") expectedTxt
+        actual <- recurseFiles $ fp [ recurseDir, "root.txt" ]
+        let
+          actualRelative = map (\f -> Path.relative recurseDir f) actual
+        Assert.equal (Set.fromFoldable expected) $ Set.fromFoldable actualRelative
+      test "one file" do
+        let
+          file = fp [ recurseDir, "c/unused.txt" ]
+
+          expected = [ file ]
+        actual <- recurseFiles file
+        Assert.equal (Set.fromFoldable expected) $ Set.fromFoldable actual
+
+{- Note to reader: Delete this line to expand comment block
+-}
+runChapterExamples :: TestSuite
+runChapterExamples = do
+  test "copyFile" do
+    let
+      inFoo = fp [ inDir, "foo.txt" ]
+
+      outFoo = fp [ outDir, "foo.txt" ]
+    copyFile inFoo outFoo
+    -- Check for valid copy
+    inFooTxt <- readTextFile UTF8 inFoo
+    outFooTxt <- readTextFile UTF8 outFoo
+    Assert.equal inFooTxt outFooTxt
+  test "getUrl" do
+    let
+      expectedOutFile = fp [ inDir, "user.txt" ]
+    str <- getUrl "https://reqres.in/api/users/1"
+    -- Check for valid read
+    expectedOutTxt <- readTextFile UTF8 expectedOutFile
+    Assert.equal expectedOutTxt str
+
+fp :: Array FilePath -> FilePath
+fp = Path.concat
```

Comment:
Yeah, `fp` is used all over the place in this file. Maybe I should rename this to `fc`.

Reply:
Two thoughts on this. First, I think using `concat` would be more readable as a whole, but then it sounds like you're concatting the elements in the array rather than concatting parts of a path. So, second, it might be better to move this function up higher in the file, so that a reader reading from top-to-bottom sees what it aliases and then can understand what's going on in the rest of the examples that follow it.
Still, `Path.concat` is likely the most readable even if not the most abbreviated.

Reply:
Went with the more readable version. `fp` was nice shorthand to use when writing these tests, but not so necessary now.